Generify ColumnReaderImpl and RecordReader (#1040)

tustvold opened the pull request:
~~This is highly experimental, I want to get further fleshing out #171 and #1037 before settling on this. In particular I want to get some numbers about performance. However, I wanted to give some visibility into what I'm doing~~

~~Builds on top of #1021~~

This introduces some limited generics into `RecordReader` and `ColumnReaderImpl` to allow for optimisations such as #1054 and #1082. Having implemented initial cuts of these, I am happy that this interface is sufficiently flexible for implementing various arrow-related optimisations.

# Which issue does this PR close?

Closes #1040.

# Rationale for this change
 
See ticket

# What changes are included in this PR?

See ticket

# Are there any user-facing changes?

No :grin:

tustvold commented on parquet/src/arrow/record_reader.rs:
```diff
@@ -16,75 +16,248 @@
 use arrow::bitmap::Bitmap;
 use arrow::buffer::{Buffer, MutableBuffer};
 
+pub(crate) mod private {
```
This is effectively my workaround for #1032 - these traits and accompanying types should not be part of the public API

tustvold commented on parquet/src/arrow/record_reader.rs:
```diff
@@ -16,75 +16,248 @@
+    ColumnValueDecoderImpl<T>,
+>;
+
+#[doc(hidden)]
```
This type is hidden from the docs, and the private module makes it impossible for users to directly construct this. The direct implication is we can finagle this type signature however we want and it not leak into downstreams

tustvold commented:
Running benchmarks on my local machine I get somewhat erratic results, from which I conclude this has no major impact on performance

```
arrow_array_reader/read Int32Array, plain encoded, mandatory, no NULLs - old                                                                             
                        time:   [3.7939 us 3.8031 us 3.8114 us]
                        change: [-3.6579% -3.4154% -3.1951%] (p = 0.00 < 0.05)
                        Performance has improved.
arrow_array_reader/read Int32Array, plain encoded, mandatory, no NULLs - new                                                                             
                        time:   [2.3030 us 2.3048 us 2.3073 us]
                        change: [+2.5908% +2.7441% +2.9142%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read Int32Array, plain encoded, optional, no NULLs - old                                                                            
                        time:   [59.193 us 59.275 us 59.363 us]
                        change: [-4.2623% -4.1285% -4.0009%] (p = 0.00 < 0.05)
                        Performance has improved.
arrow_array_reader/read Int32Array, plain encoded, optional, no NULLs - new                                                                             
                        time:   [23.209 us 23.221 us 23.236 us]
                        change: [+32.531% +32.663% +32.835%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read Int32Array, plain encoded, optional, half NULLs - old                                                                            
                        time:   [142.37 us 142.41 us 142.44 us]
                        change: [+5.5942% +6.6789% +7.7376%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read Int32Array, plain encoded, optional, half NULLs - new                                                                            
                        time:   [139.07 us 139.89 us 140.59 us]
                        change: [+0.4422% +0.9960% +1.6028%] (p = 0.00 < 0.05)
                        Change within noise threshold.
arrow_array_reader/read Int32Array, dictionary encoded, mandatory, no NULLs - old                                                                             
                        time:   [21.919 us 21.923 us 21.927 us]
                        change: [+1.3392% +1.7681% +2.0113%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read Int32Array, dictionary encoded, mandatory, no NULLs - new                                                                            
                        time:   [99.347 us 101.00 us 102.37 us]
                        change: [+5.5715% +6.7636% +8.2107%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read Int32Array, dictionary encoded, optional, no NULLs - old                                                                            
                        time:   [75.648 us 75.663 us 75.681 us]
                        change: [-1.5816% -1.5384% -1.4963%] (p = 0.00 < 0.05)
                        Performance has improved.
arrow_array_reader/read Int32Array, dictionary encoded, optional, no NULLs - new                                                                            
                        time:   [112.52 us 113.33 us 114.36 us]
                        change: [+5.2751% +7.2166% +9.0108%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read Int32Array, dictionary encoded, optional, half NULLs - old                                                                            
                        time:   [144.77 us 144.80 us 144.83 us]
                        change: [-11.013% -10.318% -9.6258%] (p = 0.00 < 0.05)
                        Performance has improved.
arrow_array_reader/read Int32Array, dictionary encoded, optional, half NULLs - new                                                                            
                        time:   [191.06 us 191.12 us 191.18 us]
                        change: [+3.4773% +3.5370% +3.5957%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read StringArray, plain encoded, mandatory, no NULLs - old                                                                            
                        time:   [800.06 us 800.19 us 800.32 us]
                        change: [-1.6826% -1.6388% -1.5967%] (p = 0.00 < 0.05)
                        Performance has improved.
arrow_array_reader/read StringArray, plain encoded, mandatory, no NULLs - new                                                                            
                        time:   [124.84 us 124.86 us 124.88 us]
                        change: [+4.1077% +4.1575% +4.2088%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read StringArray, plain encoded, optional, no NULLs - old                                                                            
                        time:   [846.35 us 846.59 us 846.87 us]
                        change: [+0.8637% +0.9228% +0.9834%] (p = 0.00 < 0.05)
                        Change within noise threshold.
arrow_array_reader/read StringArray, plain encoded, optional, no NULLs - new                                                                            
                        time:   [143.25 us 143.30 us 143.35 us]
                        change: [+2.6977% +2.7794% +2.8847%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read StringArray, plain encoded, optional, half NULLs - old                                                                            
                        time:   [773.74 us 776.61 us 779.87 us]
                        change: [+3.2218% +3.4681% +3.7063%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read StringArray, plain encoded, optional, half NULLs - new                                                                            
                        time:   [264.22 us 264.80 us 265.57 us]
                        change: [-1.3401% -1.1712% -0.9903%] (p = 0.00 < 0.05)
                        Change within noise threshold.
arrow_array_reader/read StringArray, dictionary encoded, mandatory, no NULLs - old                                                                            
                        time:   [726.17 us 726.74 us 727.44 us]
                        change: [+1.2812% +1.3725% +1.4618%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read StringArray, dictionary encoded, mandatory, no NULLs - new                                                                            
                        time:   [116.83 us 116.91 us 116.99 us]
                        change: [-3.2217% -3.0893% -2.9282%] (p = 0.00 < 0.05)
                        Performance has improved.
arrow_array_reader/read StringArray, dictionary encoded, optional, no NULLs - old                                                                            
                        time:   [802.16 us 803.89 us 805.57 us]
                        change: [-0.4055% -0.2549% -0.1073%] (p = 0.00 < 0.05)
                        Change within noise threshold.
arrow_array_reader/read StringArray, dictionary encoded, optional, no NULLs - new                                                                            
                        time:   [134.39 us 134.43 us 134.48 us]
                        change: [+0.0304% +0.2086% +0.3678%] (p = 0.02 < 0.05)
                        Change within noise threshold.
arrow_array_reader/read StringArray, dictionary encoded, optional, half NULLs - old                                                                            
                        time:   [742.00 us 742.57 us 743.00 us]
                        change: [+3.4464% +3.6453% +3.8440%] (p = 0.00 < 0.05)
                        Performance has regressed.
arrow_array_reader/read StringArray, dictionary encoded, optional, half NULLs - new                                                                            
                        time:   [236.67 us 237.14 us 238.07 us]
                        change: [+1.7094% +1.9629% +2.5264%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

What is strange to me is that this seems to have a consistent ~5% impact on the new `StringArrayReader` despite this change touching none of the code used by it. I suspect we're in the weeds of the wims of LLVM, which I'm not really sure it makes sense to optimise for at this stage - there's a lot of lower hanging fruit

yordan-pavlov commented on parquet/src/column/reader/decoder.rs:
```diff
@@ -0,0 +1,215 @@
+pub trait LevelsWriter {
+    fn capacity(&self) -> usize;
+
+    fn get(&self, idx: usize) -> i16;
```
I would expect a level writer to have a `write` method rather than a `get` method (it's a level writer, not a level getter)

tustvold replied:
This method is no longer needed - removed :+1:

yordan-pavlov commented on parquet/src/column/reader/decoder.rs:
```diff
@@ -0,0 +1,215 @@
+pub trait ColumnLevelDecoder {
+    type Writer: LevelsWriter + ?Sized;
+
+    fn create(max_level: i16, encoding: Encoding, data: ByteBufferPtr) -> Self;
```
wouldn't a `new` method be more idiomatic (instead of `create` )?

tustvold replied:
Done

yordan-pavlov commented on parquet/src/column/reader/decoder.rs:
```diff
@@ -0,0 +1,215 @@
+            if !self.decoders.contains_key(&encoding) {
+                // Initialize decoder for this page
+                let data_decoder = get_decoder::<T>(self.descr.clone(), encoding)?;
+                self.decoders.insert(encoding, data_decoder);
```
since we already have the `data_decoder` why not just return it directly after inserting, instead of looking it up a bit later?

tustvold replied:
Copy paste from old code - will fix

yordan-pavlov commented on parquet/src/column/reader/decoder.rs:
```diff
@@ -0,0 +1,215 @@
+            .current_encoding
+            .expect("current_encoding should be set");
+
+        let current_decoder = self
```
why not set a `current_decoder` field in the `set_data` method (where it has to be selected anyway to call `set_data` on it), so that it doesn't have to be looked up on every call of `read` here?

tustvold replied:
I didn't write this logic, just moved it, but my guess is this is a way to placate the borrow checker. `Decoder::get` requires a mutable reference, and we wish for decoders, in particular the dictionary decoder, to be usable across multiple `set_data` calls. 

In order to have a current_decoder construct we would either need to perform a convoluted move dance moving data in and out of the decoder map, or use `Rc<RefCell>`. This is simpler, if possibly a little less performant. FWIW I'd wager that the overheads of a hashmap keyed on a low cardinality enumeration are pretty low.

yordan-pavlov commented on parquet/src/column/reader/decoder.rs:
```diff
@@ -0,0 +1,215 @@
+
+/// An implementation of [`ColumnLevelDecoder`] for `[i16]`
+pub struct ColumnLevelDecoderImpl {
+    inner: LevelDecoderInner,
```
I wonder if the inner level decoder can be a generic parameter instead - wouldn't that remove the need to `match &mut self.inner` in the `read` method?

tustvold replied:
This would require introducing some type representation of the encoding type. This would be a fair bit of additional code/complexity that I don't think would not lead to a meaningful performance uplift. Assuming `ColumnLevelDecoderImpl::read` is called with a reasonable batch size of ~1024, the overheads of a jump table are likely to be irrelevant.

yordan-pavlov commented on parquet/src/arrow/record_reader/definition_levels.rs:
```diff
@@ -0,0 +1,82 @@
+        self.buffer.split(len)
+    }
+
+    fn writer(&mut self, batch_size: usize) -> &mut Self::Writer {
```
I wonder if `BufferSlice` would be a more fitting name to what this method does, e.g.:
```
fn writable_buffer_slice(&mut self, batch_size: usize) -> &mut Self::BufferSlice {
// return writable buffer slice
}
```

tustvold replied:
Yeah, I'm not a fan of this method's name. It gets confusing because the optimized implementations in #1054 and #1082 don't use this method to yield a slice at all, I need to think more on this

tustvold replied:
I've renamed it to `spare_capacity_mut` to reflect its similarities to the same method on `Vec`

yordan-pavlov commented on parquet/src/arrow/record_reader/definition_levels.rs:
```diff
@@ -0,0 +1,82 @@
+    ) -> impl Iterator<Item = usize> + '_ {
+        let max_def_level = self.max_level;
+        let slice = self.buffer.as_slice();
+        range.rev().filter(move |x| slice[*x] == max_def_level)
```
it might be more efficient to calculate a boolean array for the null bitmap using `arrow::compute::eq_scalar` as used in `ArrowArrayReader` here https://github.com/apache/arrow-rs/blob/master/parquet/src/arrow/arrow_array_reader.rs#L570 , because it can use SIMD (if enabled)

tustvold replied:
Currently BooleanBufferBuilder doesn't have a story for appending other BooleanBuffers - #1039 adds this but I'd rather not make this PR depend on it.

Additionally the cost of the memory allocation and copy may outweigh the gains from SIMD.

Given this I'm going to leave this as is, especially as #1054 will remove this code from the decode path for files without nested nullability.

yordan-pavlov commented on parquet/src/arrow/record_reader/buffer.rs:
```diff
@@ -0,0 +1,137 @@
+    ) {
+        let slice = self.as_slice_mut();
+
+        for (value_pos, level_pos) in range.rev().zip(rev_position_iter) {
```
it might be more efficient to insert null values using `arrow::compute::SlicesIterator` as used in `ArrowArrayReader` here https://github.com/apache/arrow-rs/blob/master/parquet/src/arrow/arrow_array_reader.rs#L570 , since it works with sequences rather than single values

tustvold replied:
This is a cool suggestion, I was not aware of this component. Unfortunately it does not appear to support reverse iteration, which is required here, so I will leave this as a potential future optimization.

tustvold commented on parquet/src/arrow/array_reader.rs:
```diff
@@ -200,7 +200,6 @@ pub struct PrimitiveArrayReader<T: DataType> {
     rep_levels_buffer: Option<Buffer>,
     column_desc: ColumnDescPtr,
     record_reader: RecordReader<T>,
-    _type_marker: PhantomData<T>,
```
This seemed to be an orphan so I just removed it

tustvold commented on parquet/src/column/reader.rs:
```diff
@@ -440,63 +435,29 @@ impl<T: DataType> ColumnReaderImpl<T> {
-    }
-
-    #[inline]
-    fn configure_dictionary(&mut self, page: Page) -> Result<bool> {
```
This logic is moved into ColumnValueDecoder

tustvold commented on parquet/src/column/reader.rs:
```diff
@@ -392,38 +419,6 @@ impl<T: DataType> ColumnReaderImpl<T> {
         Ok(true)
     }
 
-    /// Resolves and updates encoding and set decoder for the current page
```
This logic is also moved into ColumnValueDecoder

tustvold commented:
I've renamed a number of the methods and traits based on the great feedback, and also added a load of doc comments. In particular I took inspiration from std::Vec, in particular [Vec::spare_capacity_mut](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.spare_capacity_mut) and [Vec::set_len](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.set_len) which is effectively what is going on here.

I'm happy that this interface is sufficiently flexible for the optimisations I have in mind, many of which I've already got an initial cut of, and so I'm marking this ready for review.

I am aware this is a relatively complex change, to an already complex part of the codebase so if anything isn't clear please let me know.

alamb commented on parquet/src/column/reader/decoder.rs:
```diff
@@ -0,0 +1,253 @@
+use crate::util::bit_util::BitReader;
+
+/// A slice of levels buffer data that is written to by a [`ColumnLevelDecoder`]
+pub trait LevelsBufferSlice {
```
I think I missed it somewhere along the line -- what is the point of Generisizing (sp?) levels, rather than just using `[i16]`? Can definition or repetition levels ever be something other than `i16`?

tustvold replied:
Yes - #1054

alamb replied:
Ah -- got it

alamb commented on parquet/src/column/reader.rs:
```diff
@@ -440,63 +435,29 @@ impl<T: DataType> ColumnReaderImpl<T> {
+    num_buffered_values: u32,
+    encoding: Encoding,
+    buf: ByteBufferPtr,
+) -> Result<ByteBufferPtr> {
```
Is there a reason to replicate the logic in `LevelDecoder::v1(enc, max_level);` here ? Could that level decoder simply be reused? Especially since it already has tests, etc

tustvold replied:
The short answer is because I found the interface for LevelDecoder incredibly confusing, and this isn't actually interested in the decoder, just working out how many bytes of level data there are...

I can change if you feel strongly

alamb replied:
No, I was just curios

alamb reviewed (approved):
Thank you for the comments @tustvold 

I went through this code pretty carefully -- and other than the places I noted it looks like a really nice job to me. I think the additional testing such as https://github.com/apache/arrow-rs/pull/1110 gives me extra confidence that this is working as designed

To other reviewers, I would summarize this change as "pulls out common and redundant logic from some of the `RecordReader` impls into a set of common structures and traits.

alamb commented:
cc @nevi-me  @sunchao and @jorgecarleitao 

Please let us know if anyone else is interested in reviewing this PR. If not I'll plan to merge it in soon

tustvold commented on parquet/src/arrow/record_reader/buffer.rs:
```diff
@@ -0,0 +1,191 @@
+    ///
+    /// # Panics
+    ///
+    /// Implementations must panic if `len` is beyond the initialized length
```
I was trying to distinguish this from `Vec::set_len` which is unsafe because it doesn't know how much is initialized. In the case of RecordBuffer<T> the entire capacity is initialized, just possibly not set to anything useful. The result may not be desirable, but isn't UB and therefore unsafe

tustvold commented on parquet/src/arrow/record_reader/buffer.rs:
```diff
@@ -0,0 +1,191 @@
+    }
+
+    fn spare_capacity_mut(&mut self, batch_size: usize) -> &mut Self::Slice {
+        self.buffer
+            .resize((self.len + batch_size) * std::mem::size_of::<T>(), 0);
```
Sadly this is not possible with `MutableBuffer` the second parameter is u8. IMO `MutableBuffer` is pretty unfortunate and should really be typed based on what it contains, but changing this would be a major breaking change to a lot of arrow...

alamb replied:
I believe it is called `arrow2` :trollface:

tustvold replied:
Indeed `arrow2` could definitely serve as inspiration for such a change. I have some ideas on how to make such a change without major churn, but nothing fully formed just yet :grin:

jorgecarleitao replied:
`arrow2` no longer uses `MutableBuffer<T: NativeType>`: it recently migrated to `std::Vec<T: NativeType>`, for ease of use (and maintain).

tustvold replied:
> it recently migrated to std::Vec<T: NativeType>

Is there some way to force `Vec` to use stricter alignment than needed by `T`? i.e. for SIMD stuffs?

jorgecarleitao replied:
you mean e.g. use 128 bytes instead of the minimum layout required by T? I do not think it is possible on the stable channel, no.

jorgecarleitao commented on parquet/src/arrow/record_reader/buffer.rs:
```diff
@@ -0,0 +1,196 @@
+
+    #[inline]
+    pub fn as_slice(&self) -> &[T] {
+        let (prefix, buf, suffix) = unsafe { self.buffer.as_slice().align_to::<T>() };
```
Thanks @alamb for the ping. I haven't look into this PR semantics in detail because I am not familiar with this code base.

I think that this line is sound iff `T: plain old data` (in the sense that they fulfill the invariants of [Pod](https://docs.rs/bytemuck/latest/bytemuck/trait.Pod.html)).

However, `bool`, which is not Pod, [implements `ParquetValueType`](https://github.com/apache/arrow-rs/blob/master/parquet/src/data_type.rs#L652), and we pass `T: DataType::T` to TypedBuffer [here](https://github.com/apache/arrow-rs/pull/1041/files#diff-4eaa09af06edfa304fdfc1042342e7b203414904a98b48b0a11461c21643ffbeR45).

Note that like `bool`, `Int96` [contains `Option<[u32; 3]>`](https://github.com/apache/arrow-rs/blob/master/parquet/src/data_type.rs#L41) which is also not plain old data, and also [implements `ParquetValueType`](https://github.com/apache/arrow-rs/blob/master/parquet/src/data_type.rs#L796).

Maybe restrict `T` to `TypedBuffer<T: PrimitiveType>` or something, so that we do not allow non-plain old data types to be passed here?

tustvold replied:
Yeah the typing here is a bit unfortunate, there is a cludge in PrimitiveArrayReader to handle bools, and prevent Int96 but I'm not going to argue it isn't pretty gross :sweat_smile: 

It's no worse than before, but it certainly isn't ideal... I'll have a think about how to improve this without breaking the APIs :thinking:

alamb replied:
Maybe we could at least document it (or mark it as `unsafe` to force the callsites to acknowledge they aren't using `bool`)?

tustvold replied:
Going to mark this as a draft whilst I fix https://github.com/apache/arrow-rs/issues/1132 which should in turn fix this

tustvold replied:
#1155 contains the fix

tustvold commented on parquet/src/data_type.rs:
```diff
@@ -1033,21 +1032,6 @@ pub(crate) mod private {
         }
     }
-
-    /// A marker trait for [`DataType`] with a [scalar] physical type
```
This was added in #1155 but unfortunately didn't work as anticipated because of the lack of `Int16Type` which is needed for decoding levels data

alamb replied:
Can we `impl ScalarDataType for i16`?

alamb replied:
If you you need to remove this code, then we should probably reopen the original ticket https://github.com/apache/arrow-rs/issues/1132
